src/garage/torch/algos/ddpg.py

```diff
@@ -0,0 +1,300 @@
+                self.epoch_ys.append(y)
+                self.epoch_qs.append(q)
+
+        if itr % self.n_epoch_cycles == 0:
+            logger.log('Training finished')
+
+            if self.evaluate:
+                tabular.record('Epoch', epoch)
+                tabular.record('AverageReturn', np.mean(self.episode_rewards))
+                tabular.record('StdReturn', np.std(self.episode_rewards))
+                tabular.record('Policy/AveragePolicyLoss',
+                               np.mean(self.episode_policy_losses))
+                tabular.record('QFunction/AverageQFunctionLoss',
+                               np.mean(self.episode_qf_losses))
+                tabular.record('QFunction/AverageQ', np.mean(self.epoch_qs))
+                tabular.record('QFunction/MaxQ', np.max(self.epoch_qs))
+                tabular.record('QFunction/AverageAbsQ',
+                               np.mean(np.abs(self.epoch_qs)))
+                tabular.record('QFunction/AverageY', np.mean(self.epoch_ys))
+                tabular.record('QFunction/MaxY', np.max(self.epoch_ys))
+                tabular.record('QFunction/AverageAbsY',
+                               np.mean(np.abs(self.epoch_ys)))
+                if self.input_include_goal:
+                    tabular.record('AverageSuccessRate',
+                                   np.mean(self.success_history))
+
+            if not self.smooth_return:
+                self.episode_rewards = []
+                self.episode_policy_losses = []
+                self.episode_qf_losses = []
+                self.epoch_ys = []
+                self.epoch_qs = []
+
+            self.success_history.clear()
+
+        return last_average_return
+
+    def optimize_policy(self, itr, samples_data):
+        """
+        Perform algorithm optimizing.
+
+        Returns:
+            action_loss: Loss of action predicted by the policy network.
+            qval_loss: Loss of Q-value predicted by the Q-network.
+            ys: y_s.
+            qval: Q-value predicted by the Q-network.
+
+        """
+        transitions = self.replay_buffer.sample(self.buffer_batch_size)
+        transitions = self.np_to_torch(transitions)
+        observations = transitions['observation']
+        rewards = transitions['reward']
+        actions = transitions['action']
+        next_observations = transitions['next_observation']
+        terminals = transitions['terminal']
+
+        rewards = rewards.reshape(-1, 1)
+        terminals = terminals.reshape(-1, 1)
+
+        if self.input_include_goal:
+            goals = transitions['goal']
+            next_inputs = np.concatenate((next_observations, goals), axis=-1)
+            inputs = np.concatenate((observations, goals), axis=-1)
+        else:
+            next_inputs = next_observations
+            inputs = observations
+        with torch.no_grad():
+            next_actions = self.target_policy._nn_module(next_inputs)
```

Comment:
Can we not call `._nn_module` and instead just do `self.target_policy(next_inputs)`. (I assume policies are modules?)

Reply:
hmm i think there's a lot to be said for just making policies modules (is-a) vs policies having modules (has-a). why don't you try it and see what it looks like?